compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2617,6 +2626,26 @@ impl<'a> Parser<'a> {
         }
     }
 
+    fn parse_match_guard_cond(&mut self) -> PResult<'a, P<Expr>> {
+        let snapshot = self.clone();
+        let mut cond = self.parse_expr()?;
+        let (is_lhs_struct, is_rhs_struct) = match cond.kind {
+            ExprKind::Binary(_, ref lhs, ref rhs) => {
+                (matches!(lhs.kind, ExprKind::Struct(..)), matches!(rhs.kind, ExprKind::Struct(..)))
+            }
+            _ => (false, false),
+        };
```

Comment:
As you recognize here, we might want to also detect when a struct literal was parsed that instead was supposed to be an ident and the match arm body, but that is not as critical, I think.